src/test/java/build/buildfarm/cas/FileDirectoriesIndexTest.java

```diff
@@ -0,0 +1,89 @@
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package build.buildfarm.cas;
+
+import static com.google.common.truth.Truth.assertThat;
+
+import build.bazel.remote.execution.v2.Digest;
+import build.buildfarm.common.DigestUtil;
+import build.buildfarm.common.DigestUtil.HashFunction;
+import com.google.common.collect.Iterables;
+import com.google.common.jimfs.Configuration;
+import com.google.common.jimfs.Jimfs;
+import com.google.protobuf.ByteString;
+import java.io.IOException;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.util.Set;
+import org.junit.Before;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.junit.runners.JUnit4;
+
+@RunWith(JUnit4.class)
+public class FileDirectoriesIndexTest {
+  private final DigestUtil DIGEST_UTIL = new DigestUtil(HashFunction.SHA256);
+
+  private final String jdbcIndexUrl = "jdbc:sqlite:test:";
+  private Path root;
+  private FileDirectoriesIndex directoriesIndex;
+
+  @Before
+  public void setUp() throws IOException {
+    Path fileSystemRoot =
+        Iterables.getFirst(
+            Jimfs.newFileSystem(
+                    Configuration.unix()
+                        .toBuilder()
+                        .setAttributeViews("basic", "owner", "posix", "unix")
+                        .build())
+                .getRootDirectories(),
+            null);
+    root = fileSystemRoot.resolve("cache");
+    directoriesIndex = new FileDirectoriesIndex(jdbcIndexUrl, root);
+    Files.createDirectories(root);
+  }
+
+  @Test
+  public void testFileDirectoriesIndex() throws IOException {
+    // create directory and file
+    ByteString coolBlob = ByteString.copyFromUtf8("cool content");
+    Digest digest = DIGEST_UTIL.compute(coolBlob);
+    String dirName = "cool_dir";
+    Path path = root.resolve(dirName);
+    Files.write(path, coolBlob.toByteArray());
+
+    // test directoryEntries
+    Digest directory = DIGEST_UTIL.compute(coolBlob);
+    Iterable<String> entries = directoriesIndex.directoryEntries(directory);
+    for (String e : entries) {
+      assertThat(e).matches(digest.getHash());
+    }
+    // test put method
+    directoriesIndex.put(directory, entries);
+
+    // test entry-wise remove
+    for (String entry : entries) {
+      Set<Digest> digests = directoriesIndex.removeEntry(entry);
+      assertThat(digests.size() == 1 && digests.contains(directory)).isTrue();
+    }
+
+    // put again to test remove by directory
+    Files.write(path, coolBlob.toByteArray());
+    directoriesIndex.put(directory, entries);
+    directoriesIndex.remove(directory);
+    assertThat(Files.notExists(directoriesIndex.path(directory))).isTrue();
```

Comment:
This doesn't test the other side of this, whether the index thinks that it actually removed the entry index membership

Reply:
Given the previous bug, such a test should fail without this change.

Reply:
There are something i'm missing here, because you're right, the test cannot catch the previous bug, which confused me.

The bug is, (directory, entry) get inserted as (entry, directory) into the database, which means when we do SELECT/DELETE on entry, we cannot find the rows, therefore no directory get deleted. 
If that's the case, the testcase on line 79-80 should already catch it, since the returned set come from a sql SELECT on entry. If directory is used as entry, nothing should be contained in the Set.

Apparently, I'm wrong, because the test pass either way. But why?

Reply:
The directories files were removed, because they are correctly identified.

You need to assert that prior to the remove, the directories are mapped to their entries via `directoryEntries` (they will not in the case of the bug).

You need a separate test to perform `removeEntry`, and assert that the returned `Set<Digest>` contains the expected directories.

Reply:
issue found and fixed!